recitation_2/bit_operators.md

```diff
@@ -1,26 +1,428 @@
+
+So, we can filter out `myprogram.c` from our `C_FILES` with the **`filter-out`** function.
+
+- Replace `C_FILES = $(wildcard *.c)` with `C_FILES = $(filter-out $(TARGET).c, $(wildcard *.c)) `
+
+We then have to compile and link `myprogram.c` with the rest of the object files to produce `myprogram`, so we can change our rule for `myprogram` to:
+
+```bash
+$(TARGET): $(OBJS) $(TARGET).c
+	$(CC) -o $(TARGET) $(OBJS) $(TARGET).c
+```
+
+This works because the command `gcc -o <output name> <ingredients>` will both compile (if needed for `.c` files) and link your ingredients.
+
+Note: For multi-source programs, however, it is not good practice to compile and link all your files in one step because it defeats the purpose of incremental building and individual dependencies cannot be checked. In our `Makefile`, aside from the `.c` file containing the main function, all other `.o` objects are still built separately.
+
+This gives us:
+
+```bash
+CC      = gcc
+TARGET  = myprogram
+C_FILES = $(filter-out $(TARGET).c, $(wildcard *.c)) # filters out myprogram.c
+OBJS    = $(patsubst %.c,%.o,$(C_FILES))
+
+$(TARGET): $(OBJS) $(TARGET).c # compiles and links myprogram.c with rest of .o files
+	$(CC) -o $(TARGET) $(OBJS) $(TARGET).c
+	
+%.o: %.c %.h
+	$(CC) $(CFLAGS) -c -o $@ $<
+```
+
+
+
+**Adding Flags:**
+
+We want to pass compiler flags when we compile to get better debug messages to better catch errors in our code. To this, we use the flags: `-g -Wall -Werror -pedantic-errors`.
+
+To compile `hello.c` separately to `hello.c` with the flags, for example, we would do:
+
+```bash
+$ gcc -c -g -Wall -Werror -pedantic-errors -o hello.o hello.c
+```
+
+Since our `Makefile` runs shell commands for us, we can save the flags into a macro and put it in the recipe of our rule.
+
+Similarly, we might want to pass flags that enable options for linking.
+
+This gives us:
+
+```bash
+CC      = gcc
+TARGET  = myprogram
+C_FILES = $(filter-out $(TARGET).c, $(wildcard *.c))
+OBJS    = $(patsubst %.c,%.o,$(C_FILES))
+CFLAGS  = -g -Wall -Werror -pedantic-errors # compiler flags
+LDFLAGS = # linker flags
+
+$(TARGET): $(OBJS) $(TARGET).c
+	$(CC) -o $(TARGET) $(OBJS) $(TARGET).c $(LDFLAGS) # adding linker flags to recipe
+%.o: %.c %.h
+	$(CC) $(CFLAGS) -c -o $@ $< # adding compiler flags to recipe
+```
+
+**`make all` and `make clean`:**
+
+Compiling a program is not the only thing your might want to write rules for. Makefiles commonly provide instructions on a few other things besides compiling a program: 
+
+- To delete all the object files and executables so that the directory is ‘clean’. We can create a rule with the target `clean`, then with a recipe to `rm`  the object files and executables that we made.
+- An easy-to-read `all` target that is usually the first rule in the `Makefile`. This is helpful because the `make` command will build the first target in your `Makefile` , so having the first target be `all` makes it easy to see what `make` will build.
```

Comment:
```suggestion
- An easy-to-read `all` target that is usually the first rule in the `Makefile`. This is helpful because the `make` command will build the first target in your `Makefile`, so having the first target be `all` makes it easy to see what `make` will build.
```